### ISSUE TITLE
[MPS] GeLU leads to NaNs on certain BF16 inputs

### ISSUE THREAD

**Issue by Isalia20** [pytorch/pytorch#186278]
### 🐛 Describe the bug

Discovered while trying to run and compare speed between 2.12 and nightly for this model:
https://huggingface.co/nvidia/LocateAnything-3B

2.12 pytorch outputs bounding boxes correctly while nightly fails. After digging, it seems that GeLU is the reason of this. On 2.12 this:
```
import torch, torch.nn.functional as F
print(F.gelu(torch.arange(8, 16, 0.5, dtype=torch.bfloat16, device="mps"), approximate="tanh"))
```
outputs:
```
tensor([ 8.0000,  8.5000,  9.0000,  9.5000, 10.0000, 10.5000, 11.0000, 11.5000,
        12.0000, 12.5000, 13.0000, 13.5000, 14.0000, 14.5000, 15.0000, 15.5000],
       device='mps:0', dtype=torch.bfloat16)
```
On nightly:
```
tensor([ 8.0000,  8.5000,  9.0000,  9.5000, 10.0000,     nan,     nan,     nan,
            nan,     nan,     nan,     nan,     nan,     nan,     nan,     nan],
```

Highly likely this PR is the reason but haven't checked:
https://github.com/pytorch/pytorch/pull/181451

FYI @jhavukainen 

### Versions

2.12 and nightly on M5 Pro chip

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @ezyang @gchanan @kadeng @msaroufim

**Comment by jhavukainen** [comment]
@Isalia20: Yeah probably from that change. I'll see if I can repro it as well.

**Comment by jhavukainen** [comment]
Yeah the exp approximation for tanh `(exp(2x)-1)/(exp(2x)+1)` will overflow the internal fp32  a bit after the argument goes over 10 causing the inf/inf -> nan since the expression for x scales cubically. At this point however the tanh approximation is already be well saturated so we should just clamp the argument to +-10.0 without losing accuracy but prevent the overflow. I'll add a PR with the fix.
